runtime/vm/JVMImage.cpp

```diff
@@ -179,11 +193,42 @@ JVMImage::storeImageInFile()
 	}
 }
 
+extern "C" UDATA
+initializeJVMImage(J9JavaVM *vm)
+{
+    JVMImage *jvmImage = JVMImage::createInstance(vm);
+	
+	if (jvmImage->allocateImageMemory(JVMImage::INITIAL_IMAGE_SIZE) == NULL) {
+		goto _error;
+	}
+
+	if (jvmImage->isWarmRun() && jvmImage->readImageFromFile() == 0) {
+		goto _error;
+	}
+
+	if (jvmImage->initializeMonitor() == 0) {
+		goto _error;
+	}
+
+	return 1;
+
+_error:
+	shutdownJVMImage(vm);
+	return 0;
+}
+
 extern "C" void
-create_and_allocate_jvm_image(J9JavaVM *vm)
+shutdownJVMImage(J9JavaVM *vm)
 {
-	JVMImage *jvmImage = JVMImage::createInstance(vm);
-	jvmImage->allocateImageMemory(vm, JVMImage::INITIAL_IMAGE_SIZE);
+	JVMImage *jvmImage = JVMImage::getInstance();
+
+	if (jvmImage != NULL) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+
+		jvmImage->destroyMonitor();
+		jvmImage->~JVMImage();
```

Comment:
JVMImage::_jvmInstance will be set to garbage at this point. Which might cause an issue if we call JVMImage::createInstance again (Which we shouldn't call regardless). But just to be safe, we should make sure to set JVMImage::_jvmInstance to null. Since free will not null out a pointer (since it isn't a pointer to a pointer).